test/validation/api/pktio/pktio.c

```diff
@@ -2578,6 +2578,202 @@ static void pktio_test_extra_stats(void)
+	}
+	pktio_tx = pktio[0];
+	pktio_rx = (num_ifaces > 1) ? pktio[1] : pktio_tx;
+
+	CU_ASSERT_FATAL(odp_pktout_queue(pktio_tx, &pktout, 1) == 1);
+
+	ret = odp_pktio_start(pktio_tx);
+	CU_ASSERT(ret == 0);
+	if (num_ifaces > 1) {
+		ret = odp_pktio_start(pktio_rx);
+		CU_ASSERT(ret == 0);
+	}
+
+	odp_proto_stats_param_init(&param);
+	odp_proto_stats_capability(pktio_tx, &capa);
+	CU_ASSERT(capa.tx.counters.all_bits != 0);
+	param.counters.all_bits = capa.tx.counters.all_bits;
+	/* Create statistics object with all supported counters */
+	stat0 = odp_proto_stats_create("flow0_stat", &param);
+	CU_ASSERT_FATAL(stat0 != ODP_PROTO_STATS_INVALID);
+	stat1 = odp_proto_stats_create("flow1_stat", &param);
+	CU_ASSERT_FATAL(stat1 != ODP_PROTO_STATS_INVALID);
+
+	/* Flow-0 options */
+	opt0.stat = stat0;
+	opt0.oct_count0_adj = 0;
+	/* oct1 contains byte count of packets excluding Ethernet header */
+	opt0.oct_count1_adj = -14;
+
+	/* Flow-1 options */
+	opt1.stat = stat1;
+	opt1.oct_count0_adj = -8;
+	opt1.oct_count1_adj = 14;
+
+	alloc = create_packets(tx_pkt, pkt_seq, 10, pktio_tx, pktio_rx);
+
+	/* Attach statistics object to all Tx packets */
+	for (pkts = 0; pkts < alloc; pkts++) {
+		if ((pkts % 2) == 0) {
+			odp_packet_proto_stats_request(tx_pkt[pkts], &opt0);
+			flow0_pkts++;
+		} else {
+			odp_packet_proto_stats_request(tx_pkt[pkts], &opt1);
+			flow1_pkts++;
+		}
+	}
+
+	/* send */
+	for (pkts = 0; pkts != alloc; ) {
+		ret = odp_pktout_send(pktout, &tx_pkt[pkts], alloc - pkts);
+		if (ret < 0) {
+			CU_FAIL("unable to send packet\n");
+			break;
+		}
+		pkts += ret;
+	}
+	tx_pkts = pkts;
+
+	/* get */
+	for (i = 0, pkts = 0; i < 10 && pkts != tx_pkts; i++) {
+		ev = odp_schedule(NULL, wait);
+		if (ev != ODP_EVENT_INVALID) {
+			if (odp_event_type(ev) == ODP_EVENT_PACKET) {
+				pkt = odp_packet_from_event(ev);
+				if (pktio_pkt_seq(pkt) != TEST_SEQ_INVALID)
+					pkts++;
+			}
+			odp_event_free(ev);
+		}
+	}
+
+	CU_ASSERT(pkts == tx_pkts);
+
+	/* Validate Flow-0 packet statistics */
+	ret = odp_proto_stats(stat0, &data0);
+	CU_ASSERT(ret == 0);
+
+	CU_ASSERT(!(capa.tx.counters.bit.tx_pkt_drops && (data0.tx_pkt_drops > 0)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_oct_count0_drops && (data0.tx_oct_count0_drops > 0)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_oct_count1_drops && (data0.tx_oct_count1_drops > 0)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_pkts && (data0.tx_pkts != (uint64_t)flow0_pkts)));
+
+	if (capa.tx.counters.bit.tx_oct_count0) {
+		int64_t counted_bytes = PKT_LEN_NORMAL;
+
+		if (capa.tx.oct_count0_adj)
+			counted_bytes += opt0.oct_count0_adj;
+		CU_ASSERT(data0.tx_oct_count0 == counted_bytes * (uint64_t)flow0_pkts);
+	}
+
+	if (capa.tx.counters.bit.tx_oct_count1) {
+		int64_t counted_bytes = PKT_LEN_NORMAL;
+
+		if (capa.tx.oct_count1_adj)
+			counted_bytes += opt0.oct_count1_adj;
+		CU_ASSERT(data0.tx_oct_count1 == counted_bytes * (uint64_t)flow0_pkts);
+	}
+
+	/* Validate Flow-1 packet statistics */
```

Comment:
Please do not repeat the same code twice but put the common code in a function and call it for each flow.

Reply:
Thanks for suggestion, will change in next version